src/OpenTelemetry.Api/Trace/IDeferredTracerProviderBuilder.cs

```diff
@@ -0,0 +1,62 @@
+// <copyright file="IDeferredTracerProviderBuilder.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// Describes a tracer provider builder that supports deferred
+    /// initialization using an <see cref="IServiceProvider"/> to perform
+    /// dependency injection.
+    /// </summary>
+    public interface IDeferredTracerProviderBuilder
+    {
+        /// <summary>
+        /// Adds a service registration for the given instance as a <see
+        /// cref="ServiceLifetime.Singleton"/>.
+        /// </summary>
+        /// <param name="serviceType">The <see cref="Type"/> of the service.</param>
+        /// <param name="instance">The instance implementing the service.</param>
+        /// <returns>Provided <see cref="IDeferredTracerProviderBuilder"/> for chaining.</returns>
+        IDeferredTracerProviderBuilder AddService(Type serviceType, object instance);
+
+        /// <summary>
+        /// Adds a service registration using a type.
+        /// </summary>
+        /// <param name="serviceType">The <see cref="Type"/> of the service.</param>
+        /// <param name="implementationType">The <see cref="Type"/> implementing the service.</param>
+        /// <param name="lifetime">The <see cref="ServiceLifetime"/> of created instances.</param>
+        /// <returns>Provided <see cref="IDeferredTracerProviderBuilder"/> for chaining.</returns>
+        IDeferredTracerProviderBuilder AddService(Type serviceType, Type implementationType, ServiceLifetime lifetime);
+
+        /// <summary>
+        /// Adds a service registration using a factory method.
+        /// </summary>
+        /// <param name="serviceType">The <see cref="Type"/> of the service.</param>
+        /// <param name="factory">The factory used for creating service instances.</param>
+        /// <param name="lifetime">The <see cref="ServiceLifetime"/> of created instances.</param>
+        /// <returns>Provided <see cref="IDeferredTracerProviderBuilder"/> for chaining.</returns>
+        IDeferredTracerProviderBuilder AddService(Type serviceType, Func<IServiceProvider, object> factory, ServiceLifetime lifetime);
+
```

Comment:
Should we also be offering `TryAddEnumerable` and `Configure<TOptions>` for the instrumentation library authors?

Reply:
* `TryAddEnumerable` I thought about it but decided to leave it out initially. My goal was is to reinvent as little of the wheel as possible. `TryAddEnumerable` has `ServiceDescriptor` in its signature. So we have to expose a class, or add 3 (maybe more) versions of `TryAddEnumerable`. And really, what makes that method more special than the other extensions (there are lots) on `IServiceCollection`? I tried a couple of things and ultimately implemented more of the fundamental interface behind all of the extensions. Now we have `IDeferredServiceCollection` which exposes `IList<DeferredServiceDescriptor>`. That let's us do pretty much _anything_ against `IDeferredServiceCollection` that you could against `IServiceCollection`. Only downside is we're getting into a lot of code to make that work.

* `Configure<TOptions>` I don't think we really need it because all that does is register an `IConfigureOptions<T>` service which you can do any number of ways ex:

   ```csharp
   deferredServiceCollection.Add(new DeferredServiceDescriptor(
      typeof(IConfigureOptions<MyOptions>),
      typeof(MyOptionsConfiguration),
      DeferredServiceLifetime.Transient);
   ```

Reply:
Yeah with this approach of using a `IDeferredServiceCollection` offers flexibility to achieve what the extension methods do.